doc/optimizers.md

```diff
@@ -152,6 +152,170 @@ optimizer.Optimize(f, coordinates);
+  * [Decoupled Weight Decay Regularization](https://arxiv.org/abs/1711.05101)
+  * [SGD in Wikipedia](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)
+  * [SGD](#standard-sgd)
+  * [Adam: A Method for Stochastic Optimization](http://arxiv.org/abs/1412.6980)
+  * [Differentiable separable functions](#differentiable-separable-functions)
+
+## AdamW
+
+*An optimizer for [differentiable separable functions](#differentiable-separable-functions).*
+
+ AdamW is a modified version of Adam with decoupled weight decay.
+
+#### Constructors
+
+ * `AdamW()`
+ * `AdamW(`_`stepSize, batchSize`_`)`
+ * `AdamW(`_`stepSize, batchSize, weightDecay, beta1, beta2, eps, maxIterations, tolerance, shuffle`_`)`
+ * `AdamW(`_`stepSize, batchSize, weightDecay, beta1, beta2, eps, maxIterations, tolerance, shuffle, resetPolicy`_`)`
+
+
+#### Attributes
+
+| **type** | **name** | **description** | **default** |
+|----------|----------|-----------------|-------------|
+| `double` | **`stepSize`** | Step size for each iteration. | `0.001` |
+| `size_t` | **`batchSize`** | Number of points to process in a single step. | `32` |
+| `double` | **`weightDecay`** | Decay rate for the parameters. | `0.0005` |
+| `double` | **`beta1`** | Exponential decay rate for the first moment estimates. | `0.9` |
+| `double` | **`beta2`** | Exponential decay rate for the weighted infinity norm estimates. | `0.999` |
+| `double` | **`eps`** | Value used to initialize the mean squared gradient parameter. | `1e-8` |
+| `size_t` | **`max_iterations`** | Maximum number of iterations allowed (0 means no limit). | `100000` |
+| `double` | **`tolerance`** | Maximum absolute tolerance to terminate algorithm. | `1e-5` |
+| `bool` | **`shuffle`** | If true, the function order is shuffled; otherwise, each function is visited in linear order. | `true` |
+| `bool` | **`resetPolicy`** | If true, parameters are reset before every Optimize call; otherwise, their values are retained. | `true` |
+
+The attributes of the optimizer may also be modified via the member methods
+`StepSize()`, `BatchSize()`, `Beta1()`, `Beta2()`, `Eps()`, `MaxIterations()`,
+`Tolerance()`, `Shuffle()`, `WeightDecay()` and `ResetPolicy()`.
+
+#### Examples
+
+ ```c++
+ RosenbrockFunction f;
+ arma::mat coordinates = f.GetInitialPoint();
+
+ AdamW optimizer(0.01, 32, 0.0007, 0.9, 0.999, 1e-8, 100000, 1e-5, true);
+ optimizer.Optimize(f, coordinates);
+ ```
+
+#### See also:
+
+ * [Decoupled Weight Decay Regularization](https://arxiv.org/abs/1711.05101)
+ * [SGD in Wikipedia](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)
+ * [SGD](#standard-sgd)
+ * [Adam: A Method for Stochastic Optimization](http://arxiv.org/abs/1412.6980)
+ * [Differentiable separable functions](#differentiable-separable-functions)
+
+## AdamWR
+
+ *An optimizer for [differentiable separable functions](#differentiable-separable-functions).*
+
+ AdamWR is a modified version of AdamW with warm restarts and is combination of AdamW and AdamR
```

Comment:
There are a couple spaces on the start of the line here (and other places); they aren't necessary and we can remove them.  A little bit more description here could be helpful also.